# Attr data cleaning

In [1]:
import pandas as pd
import numpy as np

In [31]:
dat = pd.read_csv("attr.csv")
data = dat.copy()
pd.set_option('display.max_columns', None)

In [32]:
data.sample(5)

id  Age Attrition     BusinessTravel  DailyRate  \
536    537   60        No      Travel_Rarely       1179   
10      11   35        No      Travel_Rarely        809   
708    709   36        No         Non-Travel       1229   
796    797   25       Yes      Travel_Rarely       1219   
1015  1016   34        No  Travel_Frequently        560   

                  Department  DistanceFromHome  Education    EducationField  \
536                    Sales                16          4         Marketing   
10    Research & Development                16          3           Medical   
708                    Sales                 8          4  Technical Degree   
796   Research & Development                 4          1  Technical Degree   
1015  Research & Development                 1          4             Other   

      EmployeeCount  EmployeeNumber  EnvironmentSatisfaction Gender  \
536               1             732                        1   Male   
10                1              14                        1   Male   
708               1             990                        1   Male   
796               1            1106                        4   Male   
1015              1            1431                        4   Male   

      HourlyRate  JobInvolvement  JobLevel                JobRole  \
536           84               3         2        Sales Executive   
10            84               4         1  Laboratory Technician   
708           84               3         2        Sales Executive   
796           32               3         1  Laboratory Technician   
1015          91               3         1     Research Scientist   

      JobSatisfaction MaritalStatus  MonthlyIncome  MonthlyRate  \
536                 1        Single           5405        11924   
10                  2       Married           2426        16479   
708                 4      Divorced           5079        25952   
796                 4       Married           3691         4605   
1015                1      Divorced           2996        20284   

      NumCompaniesWorked Over18 OverTime  PercentSalaryHike  \
536                    8      Y       No                 14   
10                     0      Y       No                 13   
708                    4      Y       No                 13   
796                    1      Y      Yes                 15   
1015                   5      Y       No                 14   

      PerformanceRating  RelationshipSatisfaction  StandardHours  \
536                   3                         4             80   
10                    3                         3             80   
708                   3                         4             80   
796                   3                         2             80   
1015                  3                         3             80   

      StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
536                  0                 10                      1   
10                   1                  6                      5   
708                  2                 12                      3   
796                  1                  7                      3   
1015                 2                 10                      2   

      WorkLifeBalance  YearsAtCompany  YearsInCurrentRole  \
536                 3               2                   2   
10                  3               5                   4   
708                 3               7                   7   
796                 4               7                   7   
1015                3               4                   3   

      YearsSinceLastPromotion  YearsWithCurrManager  
536                         2                     2  
10                          0                     3  
708                         0                     7  
796                         5                     6  
1015                        1                     3

In [33]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 36 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   id                        1470 non-null   int64 
 1   Age                       1470 non-null   int64 
 2   Attrition                 1470 non-null   object
 3   BusinessTravel            1470 non-null   object
 4   DailyRate                 1470 non-null   int64 
 5   Department                1470 non-null   object
 6   DistanceFromHome          1470 non-null   int64 
 7   Education                 1470 non-null   int64 
 8   EducationField            1470 non-null   object
 9   EmployeeCount             1470 non-null   int64 
 10  EmployeeNumber            1470 non-null   int64 
 11  EnvironmentSatisfaction   1470 non-null   int64 
 12  Gender                    1470 non-null   object
 13  HourlyRate                1470 non-null   int64 
 14  JobInvolvement          

The `data.info()` allows us to see a quick overview of our data and with that we were able to deduce that the data has no missing values and it also has 36 columns and 1470 rows of data, with each column being either an integer ot an object

# Cleaning the data column by column

### Age

For this we can create a bin of ages but for that we would have to get the youngest and oldest age in the column so we know how to create the bin

In [34]:
for i in sorted(data.Age.unique()):
    print(i, end=", ")

18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 

In [35]:
print(f"The youngest age in the data is {data.Age.unique().min()}\n\
The oldest age in the data is {data.Age.unique().max()}")

The youngest age in the data is 18
The oldest age in the data is 60


In [36]:
bins = [18, 20, 30, 40, 50, 60]
group = ['<20', '20-29', '30-39', '40-49', '50-59']

# Filling NaN with 60+ because we used the right False keyword which will by default ignore the rightmost digit (60)
data['AgeGroup'] = pd.cut(data['Age'], bins=bins, labels=group, right=False).cat.add_categories('60+').fillna('60+')

# Checking the changes
data['AgeGroup'].unique()

### Attrition

In [39]:
data.Attrition.unique()

array(['Yes', 'No'], dtype=object)

Everything looks okay here so we move on to the next
### Business Travel

In [40]:
data.BusinessTravel.unique()

array(['Travel_Rarely', 'Travel_Frequently', 'Non-Travel'], dtype=object)

In [41]:
data.BusinessTravel = data.BusinessTravel.str.replace('_', ' ').str.replace("-", " ")

# Checking the changes
data.BusinessTravel.unique()

array(['Travel Rarely', 'Travel Frequently', 'Non Travel'], dtype=object)

Everything looks okay here so we move on to the next
### Department

In [42]:
data.Department.unique()

array(['Sales', 'Research & Development', 'Human Resources'], dtype=object)

Everything looks okay here so we move on to the next
### Distance From Home

In [43]:
for i in sorted(data.DistanceFromHome.unique()):
    print(i, end=', ')

1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 

In [44]:
# I will create a bin for the distance
bins = [1, 11, 20, 30]
labels = ['Very Close', 'Close', 'Far']
data['DistanceFrom_Home'] = pd.cut(data['DistanceFromHome'], bins=bins, labels=labels, right=False)

Everything looks okay here so we move on to the next
### Education

In [45]:
data.Education.unique()

array([2, 1, 4, 3, 5], dtype=int64)

In [46]:
data.Education =  data.Education.map({1: "Below College",
                                      2: "College",
                                      3: "Bachelor",
                                      4: "Master",
                                      5: "Doctor"})

In [47]:
data.Education.unique()

array(['College', 'Below College', 'Master', 'Bachelor', 'Doctor'],
      dtype=object)

Everything looks okay here so we move on to the next
### Education Field

In [48]:
data.EducationField.unique()

array(['Life Sciences', 'Other', 'Medical', 'Marketing',
       'Technical Degree', 'Human Resources'], dtype=object)

Everything looks okay here so we move on to the next
### Employee Count

In [49]:
data.EmployeeCount.unique()

array([1], dtype=int64)

Everything looks okay here so we move on to the next
### Environment Satisfaction | Job Satisfaction | Relationship Satisfaction | Job Involvement

In [50]:
columns = ['EnvironmentSatisfaction', 'RelationshipSatisfaction', 'JobInvolvement', 'JobSatisfaction']

# Confirming unique values of all four columns
print(" PRE-MAPPING\n", "-" * 11)
for i in columns:
    print(f"{i}\n{data[i].unique()}\n")

 PRE-MAPPING
 -----------
EnvironmentSatisfaction
[2 3 4 1]

RelationshipSatisfaction
[1 4 2 3]

JobInvolvement
[3 2 4 1]

JobSatisfaction
[4 2 3 1]



**They all have similarities in their unique values so I can create a function for all of them.**

In [51]:
# Define a funciton to map all four columns
def four_options(value):
    data[value] = data[value].map({1: 'Low',
                                       2: 'Medium',
                                       3: 'High',
                                       4: 'Very High'})

# For loop to implement mapping
for i in columns:
    four_options(i)
    print(f"{i} - Updated Succesfully\n")

EnvironmentSatisfaction - Updated Succesfully

RelationshipSatisfaction - Updated Succesfully

JobInvolvement - Updated Succesfully

JobSatisfaction - Updated Succesfully



In [52]:
# After mapping; check unique values again
print(" POST-MAPPING\n", "-" * 11)
for i in columns:
    print(f"{i}\n{data[i].unique()}\n")

 POST-MAPPING
 -----------
EnvironmentSatisfaction
['Medium' 'High' 'Very High' 'Low']

RelationshipSatisfaction
['Low' 'Very High' 'Medium' 'High']

JobInvolvement
['High' 'Medium' 'Very High' 'Low']

JobSatisfaction
['Very High' 'Medium' 'High' 'Low']



Everything looks okay here so we move on to the next
### Gender

In [53]:
data.Gender.unique()

array(['Female', 'Male'], dtype=object)

Everything looks okay here so we move on to the next
### Job Level | Stock Option Level

In [54]:
lvl_columns = ['JobLevel', 'StockOptionLevel']

# Confirming unique values of all four columns
print(" PRE-MAPPING\n", "-" * 11)
for i in lvl_columns:
    print(f"{i}\n{data[i].unique()}\n")

 PRE-MAPPING
 -----------
JobLevel
[2 1 3 4 5]

StockOptionLevel
[0 1 3 2]



**The purpose of this function is to make viewers understand that these columns are talking about Level so I am adding 'Lvl' to mean "Level".**

In [55]:
# Define a funciton to add level to both colunms
def level_columns(value):
    data[value] = "Lvl " + data[value].astype(str)


# Implementing the change
for i in lvl_columns:
    level_columns(i)
    print(f"{i} - Updated Succesfully\n")

JobLevel - Updated Succesfully

StockOptionLevel - Updated Succesfully



In [56]:
print(" POST-MAPPING\n", "-" * 11)
for i in lvl_columns:
    print(f"{i}\n{data[i].unique()}\n")

 POST-MAPPING
 -----------
JobLevel
['Lvl 2' 'Lvl 1' 'Lvl 3' 'Lvl 4' 'Lvl 5']

StockOptionLevel
['Lvl 0' 'Lvl 1' 'Lvl 3' 'Lvl 2']



Everything looks okay here so we move on to the next
### Job Role

In [57]:
data.JobRole.unique()

array(['Sales Executive', 'Research Scientist', 'Laboratory Technician',
       'Manufacturing Director', 'Healthcare Representative', 'Manager',
       'Sales Representative', 'Research Director', 'Human Resources'],
      dtype=object)

Everything looks okay here so we move on to the next
### Marital Status

In [58]:
data.MaritalStatus.unique()

array(['Single', 'Married', 'Divorced'], dtype=object)

Everything looks okay here so we move on to the next
### OverTime

In [59]:
data.OverTime.unique()

array(['Yes', 'No'], dtype=object)

Everything looks okay here so we move on to the next
### Performance Rating

In [60]:
data.PerformanceRating.unique()

array([3, 4], dtype=int64)

In [61]:
data.PerformanceRating = data.PerformanceRating.map({3: "Excellent",
                                                     4: "Outstanding"})

In [62]:
data.PerformanceRating.unique()

array(['Excellent', 'Outstanding'], dtype=object)

Everything looks okay here so we move on to the next
### Work-Life Balance

In [63]:
data.WorkLifeBalance.unique()

array([1, 3, 2, 4], dtype=int64)

In [64]:
data.WorkLifeBalance = data.WorkLifeBalance.map({1: 'Bad',
                                                 2: 'Good',
                                                 3: 'Better',
                                                 4: 'Best'})

In [65]:
data.WorkLifeBalance.unique()

array(['Bad', 'Better', 'Good', 'Best'], dtype=object)

Everything looks okay here so we move on to the next
### All columns that deals with years

In [66]:
# We start by creating BINS for this year group
bins = [0, 10, 20, 30]
group = ['0-10 years', '11-20 years', '21-29 years']

# Creating BINS
column = ['TotalWorkingYears', 'YearsAtCompany']

# Define a function to create the BINS
def create_bins(column):
        data[column] = pd.cut(data[column],
                               bins=bins,
                               labels=group,
                               right=False).cat.add_categories('31-40 years').fillna('31-40 years')


# Using the function created in a for loop
for i in column:
    create_bins(i)
    print(f"{i} - Updated Succesfully\n")

TotalWorkingYears - Updated Succesfully

YearsAtCompany - Updated Succesfully



In [67]:
# Appending years to the remainig data that deals with years
yrs = ['TrainingTimesLastYear', 'YearsInCurrentRole', 'YearsSinceLastPromotion', 'YearsWithCurrManager']

# Funciton to append years or year to data
def years(col):
    data[col] = data[col].astype(str) + " " + np.where(data[col] == 0, 'year', 'years')

# For loop to use the years() function
for i in yrs:
    years(i)
    print(f"{i} - Updated Succesfully\n")

TrainingTimesLastYear - Updated Succesfully

YearsInCurrentRole - Updated Succesfully

YearsSinceLastPromotion - Updated Succesfully

YearsWithCurrManager - Updated Succesfully



In [68]:
data.sample(4)

id  Age Attrition BusinessTravel  DailyRate              Department  \
1323  1324   28        No     Non Travel        280         Human Resources   
896    897   31        No  Travel Rarely       1062  Research & Development   
713    714   45        No  Travel Rarely        974  Research & Development   
167    168   30        No  Travel Rarely       1339                   Sales   

      DistanceFromHome Education EducationField  EmployeeCount  \
1323                 1   College  Life Sciences              1   
896                 24  Bachelor        Medical              1   
713                  1    Master        Medical              1   
167                  5  Bachelor  Life Sciences              1   

      EmployeeNumber EnvironmentSatisfaction  Gender  HourlyRate  \
1323            1858                    High    Male          43   
896             1252                    High  Female          96   
713              996               Very High  Female          91   
167              228                  Medium  Female          41   

     JobInvolvement JobLevel                    JobRole JobSatisfaction  \
1323           High    Lvl 1            Human Resources       Very High   
896          Medium    Lvl 2  Healthcare Representative             Low   
713            High    Lvl 1      Laboratory Technician       Very High   
167            High    Lvl 3            Sales Executive       Very High   

     MaritalStatus  MonthlyIncome  MonthlyRate  NumCompaniesWorked Over18  \
1323      Divorced           2706        10494                   1      Y   
896         Single           6812        17198                   1      Y   
713       Divorced           2270        11005                   3      Y   
167        Married           9419         8053                   2      Y   

     OverTime  PercentSalaryHike PerformanceRating RelationshipSatisfaction  \
1323       No                 15         Excellent                   Medium   
896        No                 19         Excellent                   Medium   
713        No                 14         Excellent                Very High   
167        No                 12         Excellent                     High   

      StandardHours StockOptionLevel TotalWorkingYears TrainingTimesLastYear  \
1323             80            Lvl 1        0-10 years               2 years   
896              80            Lvl 0       11-20 years               2 years   
713              80            Lvl 2        0-10 years               2 years   
167              80            Lvl 1       11-20 years               2 years   

     WorkLifeBalance YearsAtCompany YearsInCurrentRole  \
1323          Better     0-10 years            2 years   
896           Better    11-20 years            9 years   
713           Better     0-10 years            3 years   
167           Better    11-20 years            9 years   

     YearsSinceLastPromotion YearsWithCurrManager AgeGroup DistanceFrom_Home  
1323                 2 years              2 years    20-29        Very Close  
896                  1 years              8 years    30-39               Far  
713                   0 year              2 years    40-49        Very Close  
167                  7 years              4 years    30-39        Very Close

#### Selecting the columns that would bring more meaning during analysis

In [72]:
data = data[['AgeGroup', 'Gender', 'MaritalStatus', 'Attrition', 'BusinessTravel', 'DistanceFrom_Home', 'Department',
             'Education', 'EducationField', 'EnvironmentSatisfaction', 'JobInvolvement', 'JobLevel', 'JobRole',
             'JobSatisfaction', 'HourlyRate', 'DailyRate', 'MonthlyIncome', 'MonthlyRate', 'NumCompaniesWorked',
             'OverTime', 'PercentSalaryHike', 'PerformanceRating', 'RelationshipSatisfaction', 'StandardHours',
             'StockOptionLevel', 'TotalWorkingYears', 'TrainingTimesLastYear', 'WorkLifeBalance', 'YearsAtCompany', 
             'YearsInCurrentRole', 'YearsSinceLastPromotion', 'YearsWithCurrManager']]

In [70]:
# Old dataframe
dat.head(5)

id  Age Attrition     BusinessTravel  DailyRate              Department  \
0   1   41       Yes      Travel_Rarely       1102                   Sales   
1   2   49        No  Travel_Frequently        279  Research & Development   
2   3   37       Yes      Travel_Rarely       1373  Research & Development   
3   4   33        No  Travel_Frequently       1392  Research & Development   
4   5   27        No      Travel_Rarely        591  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  EmployeeNumber  \
0                 1          2  Life Sciences              1               1   
1                 8          1  Life Sciences              1               2   
2                 2          2          Other              1               4   
3                 3          4  Life Sciences              1               5   
4                 2          1        Medical              1               7   

   EnvironmentSatisfaction  Gender  HourlyRate  JobInvolvement  JobLevel  \
0                        2  Female          94               3         2   
1                        3    Male          61               2         2   
2                        4    Male          92               2         1   
3                        4  Female          56               3         1   
4                        1    Male          40               3         1   

                 JobRole  JobSatisfaction MaritalStatus  MonthlyIncome  \
0        Sales Executive                4        Single           5993   
1     Research Scientist                2       Married           5130   
2  Laboratory Technician                3        Single           2090   
3     Research Scientist                3       Married           2909   
4  Laboratory Technician                2       Married           3468   

   MonthlyRate  NumCompaniesWorked Over18 OverTime  PercentSalaryHike  \
0        19479                   8      Y      Yes                 11   
1        24907                   1      Y       No                 23   
2         2396                   6      Y      Yes                 15   
3        23159                   1      Y      Yes                 11   
4        16632                   9      Y       No                 12   

   PerformanceRating  RelationshipSatisfaction  StandardHours  \
0                  3                         1             80   
1                  4                         4             80   
2                  3                         2             80   
3                  3                         3             80   
4                  3                         4             80   

   StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
0                 0                  8                      0   
1                 1                 10                      3   
2                 0                  7                      3   
3                 0                  8                      3   
4                 1                  6                      3   

   WorkLifeBalance  YearsAtCompany  YearsInCurrentRole  \
0                1               6                   4   
1                3              10                   7   
2                3               0                   0   
3                3               8                   7   
4                3               2                   2   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     5  
1                        1                     7  
2                        0                     0  
3                        3                     0  
4                        2                     2

In [71]:
# New dataframe
data.head(5)

AgeGroup  Gender MaritalStatus Attrition     BusinessTravel  \
0    40-49  Female        Single       Yes      Travel Rarely   
1    40-49    Male       Married        No  Travel Frequently   
2    30-39    Male        Single       Yes      Travel Rarely   
3    30-39  Female       Married        No  Travel Frequently   
4    20-29    Male       Married        No      Travel Rarely   

  DistanceFrom_Home              Department      Education EducationField  \
0        Very Close                   Sales        College  Life Sciences   
1        Very Close  Research & Development  Below College  Life Sciences   
2        Very Close  Research & Development        College          Other   
3        Very Close  Research & Development         Master  Life Sciences   
4        Very Close  Research & Development  Below College        Medical   

  EnvironmentSatisfaction JobInvolvement JobLevel                JobRole  \
0                  Medium           High    Lvl 2        Sales Executive   
1                    High         Medium    Lvl 2     Research Scientist   
2               Very High         Medium    Lvl 1  Laboratory Technician   
3               Very High           High    Lvl 1     Research Scientist   
4                     Low           High    Lvl 1  Laboratory Technician   

  JobSatisfaction  HourlyRate  DailyRate  MonthlyIncome  MonthlyRate  \
0       Very High          94       1102           5993        19479   
1          Medium          61        279           5130        24907   
2            High          92       1373           2090         2396   
3            High          56       1392           2909        23159   
4          Medium          40        591           3468        16632   

   NumCompaniesWorked OverTime  PercentSalaryHike PerformanceRating  \
0                   8      Yes                 11         Excellent   
1                   1       No                 23       Outstanding   
2                   6      Yes                 15         Excellent   
3                   1      Yes                 11         Excellent   
4                   9       No                 12         Excellent   

  RelationshipSatisfaction  StandardHours StockOptionLevel TotalWorkingYears  \
0                      Low             80            Lvl 0        0-10 years   
1                Very High             80            Lvl 1       11-20 years   
2                   Medium             80            Lvl 0        0-10 years   
3                     High             80            Lvl 0        0-10 years   
4                Very High             80            Lvl 1        0-10 years   

  TrainingTimesLastYear WorkLifeBalance YearsAtCompany YearsInCurrentRole  \
0                0 year             Bad     0-10 years            4 years   
1               3 years          Better    11-20 years            7 years   
2               3 years          Better     0-10 years             0 year   
3               3 years          Better     0-10 years            7 years   
4               3 years          Better     0-10 years            2 years   

  YearsSinceLastPromotion YearsWithCurrManager  
0                  0 year              5 years  
1                 1 years              7 years  
2                  0 year               0 year  
3                 3 years               0 year  
4                 2 years              2 years

In [44]:
# Save to csv
data.to_csv("attr_cleaned.csv")

# The end